staging/src/k8s.io/apiserver/plugin/pkg/audit/webhook/webhook.go

```diff
@@ -0,0 +1,242 @@
+		buffer:       make(chan *auditinternal.Event, defaultBatchBufferSize),
+		maxBatchSize: defaultBatchMaxSize,
+		maxBatchWait: defaultBatchMaxWait,
+	}, nil
+}
+
+type batchBackend struct {
+	w *webhook.GenericWebhook
+
+	// Channel to buffer events in memory before sending them on the webhook.
+	buffer chan *auditinternal.Event
+	// Maximum number of events that can be sent at once.
+	maxBatchSize int
+	// Amount of time to wait after sending events before force sending another set.
+	//
+	// Receiving maxBatchSize events will always trigger a send, regardless of
+	// if this amount of time has been reached.
+	maxBatchWait time.Duration
+}
+
+func (b *batchBackend) Run(stopCh <-chan struct{}) error {
+	f := func() {
+		// Recover from any panics caused by this method so a panic in the
+		// goroutine can't bring down the main routine.
+		defer runtime.HandleCrash()
+
+		t := time.NewTicker(b.maxBatchWait)
+		defer t.Stop() // Release ticker resources
+
+		b.sendBatchEvents(stopCh, t.C)
+	}
+
+	go func() {
+		for {
+			f()
+		}
+	}()
+	return nil
+}
+
+// sendBatchEvents attempts to batch some number of events to the backend. It POSTs events
+// in a goroutine and logging any error encountered during the POST.
+//
+// The following things can cause sendBatchEvents to exit:
+//
+//   * Some maximum number of events are received.
+//   * Timer has passed, all queued events are sent.
+//   * StopCh is closed, all queued events are sent.
+//
+func (b *batchBackend) sendBatchEvents(stopCh <-chan struct{}, timer <-chan time.Time) {
+	var events []auditinternal.Event
+
+L:
+	for i := 0; i < b.maxBatchSize; i++ {
+		select {
+		case ev := <-b.buffer:
+			events = append(events, *ev)
+		case <-timer:
+			// Timer has expired. Send whatever events are in the queue.
+			break L
+		case <-stopCh:
+			// Webhook has shut down. Send the last events.
+			break L
+		}
+	}
+
+	if len(events) == 0 {
+		return
+	}
+
+	list := auditinternal.EventList{Items: events}
+	go func() {
+		// Execute the webhook POST in a goroutine to keep it from blocking.
+		// This lets the webhook continue to drain the queue immediatly.
+
+		defer runtime.HandleCrash()
+
+		err := webhook.WithExponentialBackoff(0, func() error {
+			return b.w.RestClient.Post().Body(&list).Do().Error()
+		})
+		if err != nil {
+			glog.Errorf("failed to POST webhook events: %v", err)
+		}
+	}()
+	return
+}
+
+var cloner = conversion.NewCloner()
+
+func (b *batchBackend) ProcessEvents(ev ...*auditinternal.Event) {
+	for _, e := range ev {
+		// Per the audit.Backend interface these events are reused after being
+		// sent to the Sink. Deep copy and send the copy to the queue.
+		event := new(auditinternal.Event)
+		if err := auditinternal.DeepCopy_audit_Event(e, event, cloner); err != nil {
+			glog.Errorf("failed to clone audit event: %v: %#v", err, e)
+			return
+		}
+
+		select {
+		case b.buffer <- event:
+		default:
+			glog.Errorf("events queue blocked, failed to send event %#v", e)
```

Comment:
I think you should just return in this case. Odds are very high that the buffer will remain blocked. The error message needs to be updated in this case though (use the loop index to log the missed events).